README.md
# Schedule Me

Software to help with the scheduling periods of time.

## Scope of this project

This project is meant to help me (and possible others) with some obstacles that will be overcome.

- Personal Time management (as I'm a workaholic)
- Flask development (I primarily work with Django...)
- Public project and colaboration (The big projects I've worked on were only with people from the places I worked on)
- maybe other obstacles, everything is welcomed.


CHANGELOG.md
# Schedule ME Changelog

## [Unreleased] - 2023-12-19

### Added

- License
- Changelog
- Design Document
- Readme
- Initial structure
- Initial User object/table
- Initial Peiods object/table

server/app/models.py
from typing import Optional
import sqlalchemy as sa
import sqlalchemy.orm as so
from sqlalchemy.sql import func
from app import db
from datetime import datetime, timezone

class User(db.Model):
    id: so.Mapped[int] = so.mapped_column(primary_key=True)
    username: so.Mapped[str] = so.mapped_column(sa.String(64), index=True, unique=True)
    email: so.Mapped[str] = so.mapped_column(sa.String(120), index=True, unique=True)
    password_hash: so.Mapped[Optional[str]] = so.mapped_column(sa.String(256))

    def __repr__(self):
        return f'<User {self.username}'

class periods(db.Model):
    id: so.Mapped[int] = so.mapped_column(primary_key=True)
    name: so.Mapped[str] = so.mapped_column(sa.String(120))
    description: so.Mapped[str] = so.mapped_column(sa.String(250))
    created_at: so.Mapped[datetime] = so.mapped_column(index=True, default=lambda: datetime.now(timezone.utc))
    start_date: so.Mapped[datetime] = so.mapped_column(sa.DateTime())
    end_date: so.Mapped[datetime] = so.mapped_column(sa.DateTime())
    user_id: so.Mapped[int] = so.mapped_column(sa.ForeignKey(User.id), index=True)
    created_by: so.Mapped[User] = so.relationship(back_populates='periods')

    def __repr__(self):
        return f'<Period {name}>'

server/app/routes.py
from app import app
from flask import render_template, flash, redirect, url_for
from app.forms import LoginForm

@app.route('/')
@app.route('/index')
def index():
    user = {'username': 'joe'}
    return render_template('base.html', title='Home', user=user)

@app.route('/myschedule')
def myschedule():
    user = {'username': 'joe'}
    periods = [
        {
        'name': 'Floating Hour',
        'description': 'Breakfeast or something',
        'starting_hour': '10',
        'starting_minute': '00',
        'ending_hour':'11',
        'ending_minute':'00',
        },
        {
        'name': 'Floating Hour',
        'description': 'Breakfeast or something',
        'starting_hour': '10',
        'starting_minute': '00',
        'ending_hour':'11',
        'ending_minute':'00',
        }
    ]
    return render_template('myschedule.html', title='My Schedule', user=user, periods=periods)

@app.route('/login', methods=['GET', 'POST'])
def login():
    form = LoginForm()
    if form.validate_on_submit():
        flash(f'Login required for user {form.username.data}, remember_me={form.remember_me.data}')
        return redirect(url_for(index))
    return render_template('login.html', title='Sign In', form=form)

server/migrations/versions/d7b9bfe1e955_.py
"""empty message

Revision ID: d7b9bfe1e955
Revises: 
Create Date: 2023-12-19 23:52:05.131745

"""
from alembic import op
import sqlalchemy as sa


# revision identifiers, used by Alembic.
revision = 'd7b9bfe1e955'
down_revision = None
branch_labels = None
depends_on = None


def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('user',
    sa.Column('id', sa.Integer(), nullable=False),
    sa.Column('username', sa.String(length=64), nullable=False),
    sa.Column('email', sa.String(length=120), nullable=False),
    sa.Column('password_hash', sa.String(length=256), nullable=True),
    sa.PrimaryKeyConstraint('id')
    )
    with op.batch_alter_table('user', schema=None) as batch_op:
        batch_op.create_index(batch_op.f('ix_user_email'), ['email'], unique=True)
        batch_op.create_index(batch_op.f('ix_user_username'), ['username'], unique=True)

    # ### end Alembic commands ###


def downgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    with op.batch_alter_table('user', schema=None) as batch_op:
        batch_op.drop_index(batch_op.f('ix_user_username'))
        batch_op.drop_index(batch_op.f('ix_user_email'))

    op.drop_table('user')
    # ### end Alembic commands ###


server/migrations/versions/00c0f5b020fe_periods_table.py
"""periods table

Revision ID: 00c0f5b020fe
Revises: d7b9bfe1e955
Create Date: 2023-12-20 00:23:54.913803

"""
from alembic import op
import sqlalchemy as sa


# revision identifiers, used by Alembic.
revision = '00c0f5b020fe'
down_revision = 'd7b9bfe1e955'
branch_labels = None
depends_on = None


def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('periods',
    sa.Column('id', sa.Integer(), nullable=False),
    sa.Column('name', sa.String(length=120), nullable=False),
    sa.Column('description', sa.String(length=250), nullable=False),
    sa.Column('created_at', sa.DateTime(), nullable=False),
    sa.Column('start_date', sa.DateTime(), nullable=False),
    sa.Column('end_date', sa.DateTime(), nullable=False),
    sa.Column('user_id', sa.Integer(), nullable=False),
    sa.ForeignKeyConstraint(['user_id'], ['user.id'], ),
    sa.PrimaryKeyConstraint('id')
    )
    with op.batch_alter_table('periods', schema=None) as batch_op:
        batch_op.create_index(batch_op.f('ix_periods_created_at'), ['created_at'], unique=False)
        batch_op.create_index(batch_op.f('ix_periods_user_id'), ['user_id'], unique=False)

    # ### end Alembic commands ###


def downgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    with op.batch_alter_table('periods', schema=None) as batch_op:
        batch_op.drop_index(batch_op.f('ix_periods_user_id'))
        batch_op.drop_index(batch_op.f('ix_periods_created_at'))

    op.drop_table('periods')
    # ### end Alembic commands ###
